lib/Amazon/S3/Thin/Credentials.pm

```diff
@@ -37,6 +52,62 @@ sub new {
     return bless $self, $class;
 }
 
+sub from_env {
+    my ($class) = @_;
+
+    # Check the environment is configured
+    croak "AWS_ACCESS_KEY_ID is not set" unless $ENV{AWS_ACCESS_KEY_ID};
+    croak "AWS_SECRET_ACCESS_KEY is not set" unless $ENV{AWS_SECRET_ACCESS_KEY};
+
+    my $self = {
+        key => $ENV{AWS_ACCESS_KEY_ID},
+        secret => $ENV{AWS_SECRET_ACCESS_KEY},
+        session_token => $ENV{AWS_SESSION_TOKEN}
+    };
+    return bless $self, $class;
+}
+
+sub from_metadata {
+    my ($class, $args) = @_;
+
+    my $ua = LWP::UserAgent->new;
+
+    # Default to the more secure v2 metadata provider
+    if (!$args->{version} or $args->{version} != 1) {
+        my $res = $ua->get('http://169.254.169.254/latest/api/token', {
+            'X-aws-ec2-metadata-token-ttl-seconds' => 90
+        });
+        croak 'Error retreiving v2 token from metadata provider: ' . $res->decoded_content
+            unless $res->is_success;
+
+        $ua->default_header('X-aws-ec2-metadata-token' => $res->decoded_content);
+    }
+
+    return _instance_metadata($ua, $args->{role});
+}
+
+sub _instance_metadata {
+    my ($ua, $role) = @_;
+
+    my $res = $ua->get('http://169.254.169.254/latest/meta-data/iam/security-credentials');
+    croak 'Error querying metadata service for roles: ' . $res->decoded_content unless $res->is_success;
+
+    my @roles = split /\n/, $res->decoded_content;
+    return unless @roles > 0;
+
+    my $target_role = (defined $role and grep { $role eq $_ } @roles)
+        ? $role
+        : $roles[0];
+
+    my $cred = $ua->get('http://169.254.169.254/latest/meta-data/iam/security-credentials/' / $target_role);
```

Comment:
This code doesn't seem to work, did you test this code on your EC2?

Reply:
Typo - sorry 😞 Fixed now!